stablehlo/tests/infer_stablehlo.mlir

```diff
@@ -1193,3 +1193,86 @@ func.func @dynamic_iota(%arg0: tensor<1xindex>) -> tensor<1xindex> {
+}
+
+// -----
+
+// CHECK-LABEL: func @rfft_with_bound
+func.func @rfft_with_bound(%arg0: tensor<?x?xf32, #stablehlo.type_extensions<bounds = [3, 10]>>) -> tensor<*xindex> {
+  %0 = "stablehlo.fft"(%arg0) {
+    fft_length = dense<9> : tensor<1xi64>, fft_type = #stablehlo<fft_type RFFT>
+  } : (tensor<?x?xf32, #stablehlo.type_extensions<bounds = [3, 10]>>) -> tensor<*xcomplex<f32>>
+  // CHECK: types0 = tensor<?x5xcomplex<f32>, #stablehlo.type_extensions<bounds = [3, ?]>>
+  %1 = "hlo_test_infer.get_return_types"(%0) : (tensor<*xcomplex<f32>>) -> tensor<*xindex>
+  func.return %1 : tensor<*xindex>
+}
+
+// -----
+
+// CHECK-LABEL: func @rfft_with_bound
+func.func @rfft_with_bound(%arg0: tensor<3x?xf32, #stablehlo.type_extensions<bounds = [?, 10]>>) -> tensor<*xindex> {
+  %0 = "stablehlo.fft"(%arg0) {
+    fft_length = dense<9> : tensor<1xi64>, fft_type = #stablehlo<fft_type RFFT>
+  } : (tensor<3x?xf32, #stablehlo.type_extensions<bounds = [?, 10]>>) -> tensor<*xcomplex<f32>>
+  // CHECK: types0 = tensor<3x5xcomplex<f32>>
+  %1 = "hlo_test_infer.get_return_types"(%0) : (tensor<*xcomplex<f32>>) -> tensor<*xindex>
+  func.return %1 : tensor<*xindex>
+}
+
+// -----
+
+// CHECK-LABEL: func @irfft_without_bound
+func.func @irfft_without_bound(%arg0: tensor<?x?xcomplex<f32>, #stablehlo.type_extensions<bounds = [3, ?]>>) -> tensor<*xindex> {
+  %0 = "stablehlo.fft"(%arg0) {
+    fft_length = dense<9> : tensor<1xi64>, fft_type = #stablehlo<fft_type IRFFT>
+  } : (tensor<?x?xcomplex<f32>, #stablehlo.type_extensions<bounds = [3, ?]>>) -> tensor<*xf32>
+  // CHECK: types0 = tensor<?x9xf32, #stablehlo.type_extensions<bounds = [3, ?]>>
+  %1 = "hlo_test_infer.get_return_types"(%0) : (tensor<*xf32>) -> tensor<*xindex>
+  func.return %1 : tensor<*xindex>
+}
+
+// -----
+
+// CHECK-LABEL: func @irfft_with_bound
+func.func @irfft_with_bound(%arg0: tensor<?x?xcomplex<f32>, #stablehlo.type_extensions<bounds = [3, 17]>>) -> tensor<*xindex> {
+  %0 = "stablehlo.fft"(%arg0) {
+    fft_length = dense<9> : tensor<1xi64>, fft_type = #stablehlo<fft_type IRFFT>
+  } : (tensor<?x?xcomplex<f32>, #stablehlo.type_extensions<bounds = [3, 17]>>) -> tensor<*xf32>
+  // CHECK: types0 = tensor<?x9xf32, #stablehlo.type_extensions<bounds = [3, ?]>>
+  %1 = "hlo_test_infer.get_return_types"(%0) : (tensor<*xf32>) -> tensor<*xindex>
+  func.return %1 : tensor<*xindex>
+}
+
+// -----
+
+func.func @irfft_invalid_bound(%arg0: tensor<3x?xcomplex<f32>, #stablehlo.type_extensions<bounds = [?, 8]>>) -> tensor<*xindex> {
+  // expected-error@+1{{IRFFT requires innermost dimension bound to be greater than or equal to fft_length[-1]. Got: 8 but fft_length is 9.}}
+  %0 = "stablehlo.fft"(%arg0) {
+    fft_length = dense<9> : tensor<1xi64>, fft_type = #stablehlo<fft_type IRFFT>
+  } : (tensor<3x?xcomplex<f32>, #stablehlo.type_extensions<bounds = [?, 8]>>) -> tensor<*xf32>
+  %1 = "hlo_test_infer.get_return_types"(%0) : (tensor<*xf32>) -> tensor<*xindex>
+  func.return %1 : tensor<*xindex>
+}
```

Comment:
3. Yes, we can. I've broken down the individual cases to be verbose, but it is not needed since `resultShape[-1] = fftLength[-1]`
4. `(C5.1) If fft_type = RFFT, dim(result, -1) = dim(operand, -1) == 0 ? 0 : dim(operand, -1) / 2 + 1.` If the shape is dynamic with bound, we would want to do the `operandBounds[-1]/2 + 1` calculation to be specific (leaving this particular description so we can refer back to the specific rule to implement in the future). But yes, for the scope of this PR, we can leave the implementation as `resultBound[-1] = ?`.

Reply:
Thank you for the update! I see that the code has been considerably simplified since the first version. Perhaps we can simplify the PR description further to follow changes in the code, along the lines of "Bounds of the result are inherited from the bounds of the operand, except for the bound of the last dimension for RFFT/IRFFT which is `?` because it is computed statically"?